src/Shared/DomainEventWrapper.cs

```diff
@@ -12,26 +13,40 @@ public DomainEventWrapper(EventGridEvent eventGridEvent)
     {
         this.Data = eventGridEvent.Data.ToObjectFromJson<JsonObject>();
         this.DomainEventName = eventGridEvent.EventType;
+        this.DomainEventSchema = EventSchema.EventGridEvent;
     }
 
-    private DomainEventWrapper(JsonObject data, string domainEventName)
+    public DomainEventWrapper(CloudEvent cloudEvent)
+    {
+        this.Data = cloudEvent.Data!.ToObjectFromJson<JsonObject>();
+        this.DomainEventName = cloudEvent.Type;
+        this.DomainEventSchema = EventSchema.CloudEvent;
+    }
+
+    private DomainEventWrapper(JsonObject data, string domainEventName, EventSchema schema)
     {
         this.Data = data;
         this.DomainEventName = domainEventName;
+        this.DomainEventSchema = schema;
     }
 
     public JsonObject Data { get; }
 
     public string DomainEventName { get; }
 
+    public EventSchema DomainEventSchema { get; }
+
     public void SetMetadata(string key, string value)
     {
-        this.Data[GetMetadataKey(key)] = value;
+        if (this.DomainEventSchema == EventSchema.EventGridEvent)
```

Comment:
Is this because metadata can only be set for EG events and not for cloud events? If so, I would throw an exception if the `EventSchema` would be `CloudEvent` considering it should be disallowed. Looking at `TryGetMetadata`, it _seems_ like EGE-specific.

Maybe we don't want to throw and opt for an alternative, to be discussed.